### PR TITLE
Split command queue into issue/completion regions and update EnqueueReadBuffer to read from completion queue

### PR DESCRIPTION
**Pull request by abhullar-tt**

### CODE REVIEW

File: tt_metal/impl/dispatch/device_command.hpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/device_command.hpp:20]: if we want it to be configurable should we allow completion size to go to 0? and in that case EnqueueReadBuffer populates the issue region. Or should there be some min size of completion region.

Comment (abhullar-tt) [reply tt_metal/impl/dispatch/device_command.hpp:20]: I prefer one data path (to device via issue and from device via completion) but a reduced issue region could lead to more stalls

Comment (DrJessop) [reply tt_metal/impl/dispatch/device_command.hpp:23]: I think this shouldn't be a constexpr value. Rather, I think there should be some defaults when constructing the command queue, and someone those defaults propagate to send kernel to device where we pass it as a compile time arg to the producer kernel.

File: tt_metal/hw/inc/dataflow_api.h
Comment (DrJessop) [inline tt_metal/hw/inc/dataflow_api.h:1550]: we should make it explicit that the rd_ptr is the completion_rd_ptr

File: tt_metal/hw/inc/dataflow_api.h
Comment (DrJessop) [inline tt_metal/hw/inc/dataflow_api.h:1560]: DeviceCommand::COMMAND_ISSUE_REGION_SIZE should be passed in as a compile-time variable.

File: tt_metal/hw/inc/dataflow_api.h
Comment (DrJessop) [inline tt_metal/hw/inc/dataflow_api.h:1572]: I added a write flush barrier that can be used here.

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/command_queue.cpp:876]: should be in terms of num pages

**Review by DrJessop — CHANGES_REQUESTED** [review]
Adding initial set of comments just so you can see them. Will continue with a second pass tomorrow.

File: tt_metal/hw/inc/dataflow_api.h
Comment (abhullar-tt) [reply tt_metal/hw/inc/dataflow_api.h:1572]: do you mean to use `noc_async_writes_flushed()` instead of `noc_async_write_barrier()`?

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp
Comment (abhullar-tt) [reply tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:325]: yeah good point

File: tt_metal/impl/dispatch/device_command.hpp
Comment (TT-billteng) [reply tt_metal/impl/dispatch/device_command.hpp:23]: how was this number calculated?

Comment (abhullar-tt) [reply tt_metal/impl/dispatch/device_command.hpp:23]: atm just arbitrarily assigned 75% of the CQ to issue region because I figured we would be doing less read backs from device

Comment (TT-billteng) [reply tt_metal/impl/dispatch/device_command.hpp:23]: can we change it to `0.75 * HUGE_PAGE_SIZE` and explain psuedo-reason behind 75%

Comment (abhullar-tt) [reply tt_metal/impl/dispatch/device_command.hpp:23]: before merging this I am planning on removing the constexpr vars representing completion/issue sizes. Instead these will be optionally user configurable. The split can be passed into `GetCommandQueue` (from https://github.com/tenstorrent-metal/tt-metal/pull/4077) 

and I will add a comment to explain the default split

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:917]: @DrJessop in this commit I resolve a hang that I was hitting because space left in completion region was less than a page. In that case host issues a wrap to completion region. 

I updated wrap index to accept one of: 
```
    enum class WrapRegion : uint8_t {
        NONE = 0,
        ISSUE = 1,
        COMPLETION = 2
    };
// defined in device_command.hpp
```

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/command_queue.cpp:724]: @DrJessop why can't we send the wrap command similar to the case for wrapping the completion? 

BTW I would like to consolidate these two paths but just pushed this to fix a hang I was hitting

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:119]: Im not sure if I should do this but there can be a case where we split our buffer read into multiple enqueue_read_buffer commands but during that process we could update our write ptr such that there isn't enough space for one page. In that case we can wrap the completion queue earlier than the next host iteration of computing how much available space there is by seeing if updated wr ptr would leave less than `min_space_required_B` in the completion queue. The completion queue read ptr is updated in same manner

Issue with this is that if the current read is the last read for the buffer and the next buffer to read has smaller page size than current, then we may have unnecessarily wrapped

If I remove this logic then host will just see there is not enough space and issue a wrap to completion (might be more straightforward to follow). Let me know your thoughts @DrJessop

File: tt_metal/hw/inc/dataflow_api.h
Comment (DrJessop) [reply tt_metal/hw/inc/dataflow_api.h]: Yes, the noc_async_writes_flushed API just ensures that the data being written has left source, but no guarantees that it has arrived to the destination.

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (DrJessop) [reply tt_metal/impl/dispatch/command_queue.cpp:917]: Shouldn't this be a function of the number of pages that you're trying to read in one shot into the completion queue (rather than just 1 padded page size)?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (DrJessop) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:119]: I'm not following. If we issue multiple enqueue read buffer commands, why can't we just have a loop that has the following:

```c++
while (not done) {
 if (wrap_condition)
  wrap();
 EnqueueReadBuffer command(...); 
 command.process()
}
```

File: tt_metal/impl/dispatch/command_queue.cpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/command_queue.cpp:917]: ah yeah so i was using the available size to figure out how big of a read to issue but I can update the logic to try and read in one shot and only issue multiple reads if buffer size is larger than completion q size

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:119]: yes will update logic

File: tt_metal/hw/inc/dataflow_api.h
Comment (abhullar-tt) [reply tt_metal/hw/inc/dataflow_api.h]: don't we want to ensure that it arrived at the dst?

File: tt_metal/tt_metal.cpp
Comment (DrJessop) [inline tt_metal/tt_metal.cpp:569]: Should check that the value is between 0 and 1.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:167]: potentially we should rename this buffer type to be "completion buffer"

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:186]: I think it would be more readable with "completion_queue_push_back"

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:129]: Potentially we should now indicate that this is a "completion queue API"

File: tt_metal/impl/dispatch/kernels/command_queue_producer.hpp
Comment (DrJessop) [inline tt_metal/impl/dispatch/kernels/command_queue_producer.hpp:54]: these would be more readable with "issue_queue_wait/pop_front"

**Review by DrJessop — CHANGES_REQUESTED** [review]
First set of changes to look over. Will maybe need another review.

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:186]: makes sense. I will rename all the `cq_*` APIs on device and host to be `issue_queue_*` and `completion_queue_*`

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:167]: makes sense because a SYSTEM_MEMORY buffer will always write into the completion queue

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:129]: `notify_host_of_cq_completion_write_pointer` -> `notify_host_of_completion_queue_write_pointer`?

File: tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp
Comment (abhullar-tt) [reply tt_metal/impl/dispatch/kernels/command_queue_consumer.hpp:167]: won't make this change because enqueue write buffer src buff type is SYSTEM_MEMORY

File: tt_metal/host_api.hpp
Comment (abhullar-tt) [inline tt_metal/host_api.hpp:64]: @TT-billteng I added the split as config to CreateDevice because in my "CQ per device" PR #4077 , the dispatch kernels are sent on device init and the issue/completion queue sizes are sent as compile time args to the dispatch kernels

File: tt_metal/hw/inc/dataflow_api.h
Comment (DrJessop) [reply tt_metal/hw/inc/dataflow_api.h]: We can keep the barrier for now and revisit.

File: tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp
Comment (DrJessop) [inline tests/tt_metal/tt_metal/unit_tests_fast_dispatch/command_queue/test_EnqueueWriteBuffer_and_EnqueueReadBuffer.cpp:412]: We should probably have tests for the extremes (where the completion queue is sized to 0% and issue queue sized to 100%)

File: tt_metal/hostdevcommon/common_runtime_address_map.h
Comment (DrJessop) [inline tt_metal/hostdevcommon/common_runtime_address_map.h:93]: was this a bug?

**Review by DrJessop — APPROVED** [review]

File: tt_metal/hostdevcommon/common_runtime_address_map.h
Comment (abhullar-tt) [reply tt_metal/hostdevcommon/common_runtime_address_map.h:93]: No, I added src and dst page offsets to support multiple read/write commands for one buffer

File: tt_metal/host_api.hpp
Comment (TT-billteng) [reply tt_metal/host_api.hpp:64]: who would be the right audience that would want to change this param? I feel like this is too much detail in an API as often used as `CreateDevice`

**Review by TT-billteng — APPROVED** [review]
Looks great :)

**Pull request merged**
